### PR TITLE
rkt support

### PR DESCRIPTION
**Pull request by bcwaldon**

### CODE REVIEW

**Comment by bcwaldon** [comment]
@yifan-gu I believe all copyrights should be assigned to CoreOS, Inc in 2015

File: Godeps/Godeps.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
 	"ImportPath": "github.com/GoogleCloudPlatform/kubernetes",
-	"GoVersion": "go1.4",
+	"GoVersion": "go1.4.1",
```
Comment (bcwaldon) [inline Godeps/Godeps.json:3]: Revert this change.

File: pkg/kubelet/runonce.go [hunk 2]
```diff
@@ -88,34 +88,36 @@ func (kl *Kubelet) runOnce(pods []api.BoundPod) (results []RunPodResult, err err
...
-		if retry >= RunOnceMaxRetries {
-			return fmt.Errorf("timeout error: pod %q containers not running after %d retries", pod.Name, RunOnceMaxRetries)
-		}
-		// TODO(proppy): health checking would be better than waiting + checking the state at the next iteration.
-		glog.Infof("pod %q containers synced, waiting for %v", pod.Name, delay)
-		<-time.After(delay)
-		retry++
-		delay *= RunOnceRetryDelayBackoff
-	}
+	return fmt.Errorf("not implemented")
```
Comment (bcwaldon) [inline pkg/kubelet/runonce.go:91]: This is clearly not going to be ok upstream. What needs to happen here?

File: pkg/kubelet/rocket/rocket.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+	}
+	defer unitFile.Close()
+
+	_, err = io.Copy(unitFile, unit.Serialize(units))
+	if err != nil {
+		return "", err
+	}
+
+	// Enable unit file.
+	_, _, err = r.systemd.EnableUnitFiles([]string{path.Join(tmpDirPath, name)}, true, true)
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:243]: Why are you enabling the unit files?

File: pkg/kubelet/rocket/rocket.go [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+	}
+	uuid := strings.TrimSpace(string(output))
+	if uuid == "" {
+		panic("expect uuid returned, but get nothing")
+	}
+	fmt.Println("Prepare uuid:", uuid)
+
+	units = append(units,
+		&unit.UnitOption{
+			Section: "X-K8S",
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:218]: Can we please spell out Kubernetes?

File: pkg/kubelet/rocket/rocket.go [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+	for _, c := range pod.Spec.Containers {
+		images = append(images, c.Image)
+	}
+	output, err := r.runCommand(true, "prepare", images...)
+	if err != nil {
+		return "", err
+	}
+	uuid := strings.TrimSpace(string(output))
+	if uuid == "" {
+		panic("expect uuid returned, but get nothing")
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:212]: This can be an error, right?

File: pkg/kubelet/rocket/rocket.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+		{
+			Section: "Unit",
+			Name:    "Description",
+			Value:   "k8s_rocket_pod",
+		},
+
+		{
+			Section: "Install",
+			Name:    "WantedBy",
+			Value:   "multi-user.target", // TODO(yifan): Hardcode.
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:86]: Why is this option necessary?

File: pkg/kubelet/rocket/rocket.go [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return &RocketRuntime{
+		systemd:  systemd,
+		endpoint: endpoint,
+	}, nil
+}
+
+func (r *RocketRuntime) runCommand(stdout bool, subCommand string, args ...string) ([]byte, error) {
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:54]: The `stdout` argument is always set to `true`. Just remove this flag altogether.

File: api/examples/pod.json [hunk 8]
```diff
@@ -8,11 +8,16 @@
       "id": "php",
       "containers": [{
         "name": "nginx",
-        "image": "dockerfile/nginx",
+        "image": "/home/yifan/gopher/src/playground/hello1.aci",
```
Comment (bcwaldon) [inline api/examples/pod.json:11]: Changes to this file need to be reverted.

File: pkg/kubelet/container.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	"hash/adler32"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/rocket"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+)
+
+const defaultEndpoint = "/home/yifan/gopher/src/github.com/coreos/rocket/bin/rkt"
```
Comment (bcwaldon) [inline pkg/kubelet/container.go:29]: 1. This is not an "endpoint", its a path to a binary/executable.
2. We clearly can't use `/home/yifan/...` here. Can we rely on $PATH to find `rkt`?

File: pkg/kubelet/container.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+func init() {
+	var err error
+	rkt, err = rocket.NewRocketRuntime(defaultEndpoint)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// FindContainerInPod finds the container in the pod.
```
Comment (bcwaldon) [inline pkg/kubelet/container.go:41]: This style of docstring that just repeats what the method is named is really just white noise. Either beef up the description or drop it altogether.

File: pkg/kubelet/container.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+	// Update the pod and start it.
+	pod.Spec.Containers = append(pod.Spec.Containers, *container)
+	boundPod := &api.BoundPod{pod.TypeMeta, pod.ObjectMeta, pod.Spec}
+	if err := rkt.RunPod(boundPod); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Help function, remove the container in the given pod.
```
Comment (bcwaldon) [inline pkg/kubelet/container.go:67]: The dosctring convention of starting with the method name applies to unexported functions, too.

File: pkg/kubelet/rocket/rocket.go [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+
+func (r *RocketRuntime) KillPod(pod *api.Pod) error {
+	glog.V(4).Infof("Killing pod: name %s", pod.Name)
+	name := fmt.Sprintf("K8S_%s_%s.service", pod.Name, pod.Namespace)
+
+	_, err := r.systemd.DisableUnitFiles([]string{path.Join(tmpDirPath, name)}, true)
+	if err != nil {
+		return err
+	}
+	r.systemd.KillUnit(name, 9)
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:301]: Please assign `9` to a constant to add a bit more context.

File: pkg/kubelet/rocket/rocket.go [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+		},
+		&unit.UnitOption{
+			Section: "Service",
+			Name:    "ExecStart",
+			Value:   r.constructRunCmd(uuid),
+		},
+	)
+
+	// Save the unit file.
+	name := fmt.Sprintf("K8S_%s_%s.service", pod.Name, pod.Namespace)
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:230]: This name construction is done in two places. Move it out into a helper function

File: pkg/kubelet/rocket/rocket.go [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+	allArgs = append(allArgs, subCommand)
+	allArgs = append(allArgs, args...)
+	cmd := exec.Command(r.endpoint, allArgs...)
+	if !stdout {
+		return nil, cmd.Start()
+	}
+	return cmd.Output()
+}
+
+func (r *RocketRuntime) constructRunCmd(uuid string) string {
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:66]: This is only used once. Just replace the calling code with `fmt.Sprintf("%s run-prepared %s", r.endpoint, uuid)`

File: pkg/kubelet/rocket/rocket.go [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+	b, err := json.Marshal(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	units := []*unit.UnitOption{
+		{
+			Section: "Unit",
+			Name:    "Description",
+			Value:   "k8s_rocket_pod",
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:80]: This description is meant for human consumption. Either add some real information here or just drop `Description` altogether.

File: pkg/kubelet/rocket/rocket.go [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+		}
+		podState = pods[rktID].State
+	}
+
+	fmt.Println("pod state", rktID, podState)
+
+	pod.Status.Info = make(map[string]api.ContainerStatus)
+	for _, container := range pod.Spec.Containers {
+		switch podState {
+		case "running":
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:170]: Are these strings we can pull from constant values elsewhere?

File: pkg/kubelet/rocket/rocket.go [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+				},
+			}
+		case "exited", "deleting", "gone":
+			pod.Status.Info[container.Name] = api.ContainerStatus{
+				State: api.ContainerState{
+					Termination: &api.ContainerStateTerminated{},
+				},
+			}
+		default:
+			panic("Unexpected state")
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:189]: Again with the panic. Why panic rather than return an error?

File: pkg/kubelet/rocket/rocket.go [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+
+	var pod api.Pod
+	opts, err := unit.Deserialize(f)
+	if err != nil {
+		return nil, err
+	}
+
+	var rktID string
+	for _, opt := range opts {
+		if opt.Section != "X-K8S" {
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:139]: Please use constants for any strings like "X-K8S" or "RocketID" or "POD" that are used in multiple places.

File: pkg/kubelet/rocket/rocket.go [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+		&unit.UnitOption{
+			Section: "Service",
+			Name:    "ExecStart",
+			Value:   r.constructRunCmd(uuid),
+		},
+	)
+
+	// Save the unit file.
+	name := fmt.Sprintf("K8S_%s_%s.service", pod.Name, pod.Namespace)
+	unitFile, err := os.Create(path.Join(tmpDirPath, name))
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:231]: Use the builtin http://golang.org/pkg/io/ioutil/#TempFile for management of temporary files like this.

File: pkg/kubelet/rocket/rocket.go [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+	output, err := r.runCommand(true, "list", "--no-legend")
+	glog.V(4).Infof("list output: %s", string(output))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(output) == 0 { // No containers running.
+		return nil, nil
+	}
+	for _, line := range strings.Split(strings.TrimSpace(string(output)), "\n") {
```
Comment (bcwaldon) [inline pkg/kubelet/rocket/rocket.go:112]: Whenever you're parsing output like this it helps to document what the data should look like. Just dump an example of `rkt list --no-legend` somewhere around here.

File: pkg/kubelet/container.go [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+func removeContainer(pod *api.Pod, container api.Container) {
```
Comment (bcwaldon) [inline pkg/kubelet/container.go:68]: This method signature is the only one that swaps the ordering of `pod` and `container`. Is there a good reason for that?

**Comment by bcwaldon** [comment]
@yifan-gu What's the plan for testing all of this new code?

**Comment by yifan-gu** [comment]
@bcwaldon Thanks for the comments and sorry about the messy code, just a fast implementation, will ping you after clean up.

For testing, I guess we can have a fakeRocketClient, sort like the fakeDockerClient stuff?

**Comment by bcwaldon** [comment]
@yifan-gu that would make sense

**Comment by bcwaldon** [comment]
@yifan-gu or we punt on fakeXClient and provide a fake at the ContainerRuntimeInterface layer that abstracts the container system away altogether.

**Comment by yifan-gu** [comment]
@bcwaldon Totally agree. 
I think we also need to test the rocket's implementation of the container runtime interface. Just do some simple creation/deletion/listing on predefined images could be fine.

**Comment by yifan-gu** [comment]
@bcwaldon moved to #4, feel free to reopen if necessary!
